src/ext/legit.lisp

```diff
@@ -0,0 +1,507 @@
+(define-key lem/peek-legit::*peek-legit-keymap* "M-p" 'legit-previous-header)
+(define-key *legit-diff-mode-keymap* "Tab" 'other-window)
+
+;; help
+(define-key lem/peek-legit::*peek-legit-keymap* "?" 'legit-help)
+(define-key lem/peek-legit::*peek-legit-keymap* "C-x ?" 'legit-help)
+;; quit
+(define-key *legit-diff-mode-keymap* "q" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "q" 'legit-quit)
+(define-key *legit-diff-mode-keymap* "M-q" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "M-q" 'peek-legit-quit)
+(define-key *legit-diff-mode-keymap* "Escape" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "Escape" 'peek-legit-quit)
+(define-key *legit-diff-mode-keymap* "C-c C-k" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "C-c C-k" 'legit-quit)
+
+(defun pop-up-message (message)
+  (with-pop-up-typeout-window (s (make-buffer "*legit status*") :erase t)
+    (format s "~a" message)))
+
+(defun last-character (s)
+  (subseq s (- (length s) 2) (- (length s) 1)))
+
+(defmacro with-current-project (&body body)
+  "Execute body with the current working directory changed to the project's root,
+  find and set the VCS system for this operation.
+
+  If no Git directory (or .fossil file) are found, message the user."
+  `(let ((root (lem-core/commands/project:find-root (buffer-directory))))
+     (uiop:with-current-directory (root)
+       (multiple-value-bind (root vcs)
+           (porcelain::vcs-project-p)
+         (if root
+             (let ((porcelain::*vcs* vcs))
+               (progn
+                 ,@body))
+             (message "Not inside a version-controlled project?"))))))
+
+
+;;; Git commands
+;;; that operate on files.
+;;;
+;;; Global commands like commit need not be defined here.
+
+;; diff
+(defun show-diff (diff)
+  ;; Show a diff in the *legit-diff* buffer.
+  ;; Share usage between showing file diff ("move" function)
+  ;; and showing a commit.
+  (let ((buffer (lem-base:get-or-create-buffer "*legit-diff*")))
+    (setf (buffer-directory buffer)
+          (uiop:getcwd))
+    (setf (buffer-read-only-p buffer) nil)
+    (erase-buffer buffer)
+    (move-to-line (buffer-point buffer) 1)
+    (insert-string (buffer-point buffer) diff)
+    (change-buffer-mode buffer 'legit-diff-mode)
+    (setf (buffer-read-only-p buffer) t)
+    (move-to-line (buffer-point buffer) 1)))
+
+(defun make-move-function (file  &key cached)
+  (lambda ()
+    (with-current-project ()
+      (show-diff (porcelain::file-diff file :cached cached)))))
+
+(defun make-visit-file-function (file)
+  ;; note: the lambda inside the loop is not enough, it captures the last loop value.
+  (lambda ()
+    file))
+
+;; show commit.
+(defun make-show-commit-function (ref)
+  (lambda ()
+    (with-current-project ()
+      (show-diff (porcelain::show-commit-diff ref)))))
+
+;; stage
+(defun make-stage-function (file)
+  (lambda ()
+    (with-current-project ()
+      (porcelain::stage file)
```

Comment:
Is it possible to export each symbol in the porcelain package?
https://google.github.io/styleguide/lispguide.xml?showone=Omit_library_prefixes#Omit_library_prefixes